docs/getting-started/walkthroughs/datastream_api.md

```diff
@@ -0,0 +1,897 @@
+
+<p class="text-center">
+    <img alt="Transactions" width="80%" src="{{ site.baseurl }}/fig/fraud-transactions.svg"/>
+</p>
+
+Transactions 3 and 4 should be marked as fraudulent because it is a small transaction, $0.09, followed by a large one, $510.
+Alternatively, transactions 7, 8, and 9 are not fraud because the small amount of $0.02 is not immediately followed by the large one; instead, there is an intermediate transaction that breaks the pattern.
+
+To do this, the fraud detector must _remember_ information across events; a large transaction is only fraudulent if the previous one was small.
+Remembering information across events requires [state]({{ site.baseurl }}/concepts/glossary.html#managed-state) and so you will implement your fraud detector using a [KeyedProcessFunction]({{ site.baseurl }}/dev/stream/operators/process_function.html), which provides fine-grained control over state and time.
+
+The most straightforward implementation would be to set a flag whenever a small transaction is processed.
+This way, when a large transaction comes through, you can check if the flag is set for that account.
+If it is, then this is fraud and output an alert.
+This flag is what you want to store in Flink state.
+
+The most basic type of state in Flink is [ValueState]({{ site.baseurl }}/dev/stream/state/state.html#using-managed-keyed-state), a data type that provides _fault tolerant_, _managed_, _per key_ state.
+`ValueState` is created using a `ValueStateDescriptor` which contains metadata about how it should be managed.
+
+<div class="codetabs" markdown="1">
+<div data-lang="java" markdown="1">
+{% highlight java %}
+public class FraudDetector extends KeyedProcessFunction<Long, Transaction, Alert> {
+
+	public static final double SMALL_AMOUNT = 0.01;
+
+	public static final double LARGE_AMOUNT = 500.00;
+
+	public static final long ONE_DAY = 24 * 60 * 60 * 1000;
+
+	private transient ValueState<Boolean> flagState;
+
+	@Override
+	public void open(Configuration parameters) throws Exception {
+		ValueStateDescriptor<Boolean> flagDescriptor = new ValueStateDescriptor<>(
+				"flag",
+				Types.BOOLEAN);
+
+		flagState = getRuntimeContext().getState(flagDescriptor);
+	}
+{% endhighlight %}
+</div>
+
+<div data-lang="scala" markdown="1">
+{% highlight scala %}
+class FraudDetector extends KeyedProcessFunction[Long, Transaction, Alert] {
+
+    @transient private var flagState: ValueState[java.lang.Boolean] = _
+
+    @throws[Exception]
+    override def open(parameters: Configuration): Unit = {
+        val flagDescriptor = new ValueStateDescriptor("flag", Types.BOOLEAN)
+        flagState = getRuntimeContext.getState(flagDescriptor)
+    }
+{% endhighlight %}
+</div>
+</div>
+
+`ValueState` is a wrapper class, similar to `AtomicReference` in the Java standard library.
+It provides three methods for interacting with its contents; `update` sets the state, `value` gets the current value, and `clear` to delete its contents.
```

Comment:
Especially with the example of being similar to `AtomicReference`, we should clarify that this may be a copy but may also be a reference (depending on the state back-end, but don't include that - it would go too far). Any changes to object returned by `value()` are only guaranteed to be reflected in state after calling `update`!

Reply:
I did, but it was further down. I reworked it to be in this paragraph.